src/workspace/mod.rs

```diff
@@ -0,0 +1,1823 @@
+          config_file: c.clone(),
+        })
+      })
+      .collect()
+  }
+
+  pub fn jsr_packages_for_publish(&self) -> Vec<JsrPackageConfig> {
+    let ctx = self.resolve_start_ctx();
+    let Some(config) = &ctx.deno_json else {
+      return Vec::new();
+    };
+    let deno_json = &config.member;
+    if deno_json.dir_path() == self.root_dir.to_file_path().unwrap() {
+      return self.jsr_packages();
+    }
+    match ctx.maybe_package_config() {
+      Some(pkg) => vec![pkg],
+      None => Vec::new(),
+    }
+  }
+
+  pub fn npm_packages(&self) -> Vec<NpmPackageConfig> {
+    self
+      .package_jsons()
+      .filter_map(|c| {
+        Some(NpmPackageConfig {
+          member_ctx: self.resolve_member_ctx(&c.specifier()),
+          package_nv: PackageNv {
+            name: c.name.clone()?,
+            version: {
+              let version = c.version.as_ref()?;
+              deno_semver::Version::parse_from_npm(version).ok()?
+            },
+          },
+          package_json: c.clone(),
+        })
+      })
+      .collect()
+  }
+
+  pub fn resolve_deno_json(
+    &self,
+    specifier: &Url,
+  ) -> Option<(&UrlRc, &ConfigFileRc)> {
+    self.resolve_deno_json_str(specifier.as_str())
+  }
+
+  fn resolve_deno_json_str(
+    &self,
+    specifier: &str,
+  ) -> Option<(&UrlRc, &ConfigFileRc)> {
+    let mut specifier = specifier;
+    if !specifier.ends_with('/') {
+      specifier = parent_specifier_str(specifier)?;
+    }
+    loop {
+      let (folder_url, folder) = self.resolve_folder_str(specifier)?;
+      if let Some(config) = folder.deno_json.as_ref() {
+        return Some((folder_url, config));
+      }
+      specifier = parent_specifier_str(folder_url.as_str())?;
+    }
+  }
+
+  fn resolve_pkg_json_str(
+    &self,
+    specifier: &str,
+  ) -> Option<(&UrlRc, &PackageJsonRc)> {
+    let mut specifier = specifier;
+    if !specifier.ends_with('/') {
+      specifier = parent_specifier_str(specifier)?;
+    }
+    loop {
+      let (folder_url, folder) = self.resolve_folder_str(specifier)?;
+      if let Some(pkg_json) = folder.pkg_json.as_ref() {
+        return Some((folder_url, pkg_json));
+      }
+      specifier = parent_specifier_str(folder_url.as_str())?;
+    }
+  }
+
+  pub fn root_folder(&self) -> (&UrlRc, &FolderConfigs) {
+    (
+      &self.root_dir,
+      self.config_folders.get(&self.root_dir).unwrap(),
+    )
+  }
+
+  pub fn config_folders(&self) -> &IndexMap<UrlRc, FolderConfigs> {
+    &self.config_folders
+  }
+
+  pub fn resolve_folder(
+    &self,
+    specifier: &Url,
+  ) -> Option<(&UrlRc, &FolderConfigs)> {
+    self.resolve_folder_str(specifier.as_str())
+  }
+
+  fn resolve_folder_str(
+    &self,
+    specifier: &str,
+  ) -> Option<(&UrlRc, &FolderConfigs)> {
+    let mut best_match: Option<(&UrlRc, &FolderConfigs)> = None;
+    // it would be nice if we could store this config_folders relative to
+    // the root, but the members might appear outside the root folder
```

Comment:
Err, I'm confused here - I thought the workspace members must be descendants of the workspace root. Isn't that the case?

Reply:
Yeah, this an old comment. I will fix it.